runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -29,23 +30,13 @@ import kotlin.coroutines.coroutineContext
  * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
  * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
  * after the entire body has been streamed.
- *
- * @param checksumAlgorithmNameInitializer a function which parses the input [I] to return the checksum algorithm name
  */
 @InternalApi
-public class FlexibleChecksumsRequestInterceptor<I>(
-    private val checksumAlgorithmNameInitializer: (I) -> String?,
-) : HttpInterceptor {
-    private var checksumAlgorithmName: String? = null
-
-    override fun readAfterSerialization(context: ProtocolRequestInterceptorContext<Any, HttpRequest>) {
-        @Suppress("UNCHECKED_CAST")
-        val input = context.request as I
-        checksumAlgorithmName = checksumAlgorithmNameInitializer(input)
-    }
+public class FlexibleChecksumsRequestInterceptor : HttpInterceptor {
+    override suspend fun modifyBeforeSigning(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
+        val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor>()
 
-    override suspend fun modifyBeforeRetryLoop(context: ProtocolRequestInterceptorContext<Any, HttpRequest>): HttpRequest {
-        val logger = coroutineContext.logger<FlexibleChecksumsRequestInterceptor<I>>()
+        val checksumAlgorithmName = context.executionContext.getOrNull(HttpOperationContext.ChecksumAlgorithm)
```

Comment:
Yeah, I wanted to have a discussion about these interceptor changes + how they affect binary compatibility. I'll take a look at the second option since it seems better

Reply:
I ended up going with the first option because the `FlexibleChecksumsResponseInterceptor` still takes the input type param `<I>`